Sources/SourceKitLSP/Swift/SwiftCodeLensScanner.swift

```diff
@@ -0,0 +1,86 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LanguageServerProtocol
+import SwiftSyntax
+
+/// Scans a source file for classes or structs annotated with `@main` and returns a code lens for them.
+final class SwiftCodeLensScanner: SyntaxVisitor {
+  /// The document snapshot of the syntax tree that is being walked.
+  private let snapshot: DocumentSnapshot
+
+  /// The collection of CodeLenses found in the document.
+  private var result: [CodeLens] = []
+
+  /// The map of supported commands and their client side command names
+  private let supportedCommands: [String: String]
+
+  private init(snapshot: DocumentSnapshot, supportedCommands: [String: String]) {
+    self.snapshot = snapshot
+    self.supportedCommands = supportedCommands
+    super.init(viewMode: .fixedUp)
+  }
+
+  /// Public entry point. Scans the syntax tree of the given snapshot for an `@main` annotation
+  /// and returns CodeLens's with Commands to run/debug the application.
+  public static func findCodeLenses(
+    in snapshot: DocumentSnapshot,
+    syntaxTreeManager: SyntaxTreeManager,
+    supportedCommands: [String: String]
+  ) async -> [CodeLens] {
+    guard snapshot.text.contains("@main") && !supportedCommands.isEmpty else {
+      // This is intended to filter out files that obviously do not contain an entry point.
+      return []
+    }
+
+    let syntaxTree = await syntaxTreeManager.syntaxTree(for: snapshot)
+    let visitor = SwiftCodeLensScanner(snapshot: snapshot, supportedCommands: supportedCommands)
+    visitor.walk(syntaxTree)
+    return visitor.result
+  }
+
+  override func visit(_ node: ClassDeclSyntax) -> SyntaxVisitorContinueKind {
+    node.attributes.forEach(self.captureLensFromAttribute)
+    return .skipChildren
+  }
+
+  override func visit(_ node: StructDeclSyntax) -> SyntaxVisitorContinueKind {
+    node.attributes.forEach(self.captureLensFromAttribute)
+    return .skipChildren
+  }
+
+  private func captureLensFromAttribute(attribute: AttributeListSyntax.Element) {
+    if attribute.trimmedDescription == "@main" {
+      let range = self.snapshot.absolutePositionRange(of: attribute.trimmedRange)
+
+      if let runCommand = supportedCommands["swift.run"] {
```

Comment:
I think it would be good to have some kind of documentation about which commands SourceKit-LSP knows about. I don’t have strong opinions on where these go. As we add more, it might get difficult for people implementing a new editor to find the commands they can add support for.

Options I can think of are:
- Adding it to the doc comment on `CodeLensOptions.supportedComamands`
- Creating an enum with the known commands and then use the raw value of the as keys to `supportedCommands`
- Creating a document in the Documentation folder

Reply:
What happens if a client sends an unknown supported command in the client capabilities of the initialize request (I know that VS Code can’t do that and we use the workaround via `initializationOptions` but other editors might)? I think the initialize request would fail to decode and SourceKit-LSP wouldn’t start. And that’s a likely scenario because the client should be able to announce capabilities eg. supported by an upcoming version of SourceKit-LSP.

It would probably be to use a struct with static members (like `BuildTargetTag` was the first example I found). We use that pattern in a few places in SourceKit-LSP.

Reply:
The supported commands [aren't actually decoded or encoded in the code lens client capabilities](https://github.com/swiftlang/sourcekit-lsp/pull/1556/files#diff-51fa3bcba74e122b6770fc1d2462a5d27796e6af1ad5a6dfb636f54fc6437981R486); they can only be set currently via `initializationOptions`. Is initializationOptions a workaround for VS Code and supportedCommands should be encodable/decodable so they can be set in the initialize request proper?

Reply:
> The supported commands [aren't actually decoded or encoded in the code lens client capabilities](https://github.com/swiftlang/sourcekit-lsp/pull/1556/files#diff-51fa3bcba74e122b6770fc1d2462a5d27796e6af1ad5a6dfb636f54fc6437981R486)

Oh, I missed that. We should be able to decode them.

> Is initializationOptions a workaround for VS Code and supportedCommands should be encodable/decodable so they can be set in the initialize request proper

Yes

Reply:
OK, latest push should address this.